vslib/src/SwitchStateBaseMACsec.cpp

```diff
@@ -0,0 +1,730 @@
+    // Find ACL Entry
+    attr.id = SAI_ACL_ENTRY_ATTR_ACTION_MACSEC_FLOW;
+    attr.value.aclaction.parameter.oid = macsec_flow_id;
+    attr.value.aclaction.enable = true;
+    std::vector<sai_object_id_t> acl_entry_ids;
+    findObjects(SAI_OBJECT_TYPE_ACL_ENTRY, attr, acl_entry_ids);
+    if (acl_entry_ids.empty())
+    {
+        SWSS_LOG_DEBUG(
+            "Cannot find corresponding ACL entry for the flow %s",
+            sai_serialize_object_id(macsec_flow_id).c_str());
+        return SAI_STATUS_FAILURE;
+    }
+    // Find ACL Table
+    sai_object_id_t acl_table_id = SAI_NULL_OBJECT_ID;
+    // All entries with same MACsec flow correspond one ACL Table
+    // Because an ACL Table corresponds one MACsec Port
+    // And a flow never belongs to two MACsec Port
+    if (getACLTable(acl_entry_ids.front(), acl_table_id) != SAI_STATUS_SUCCESS || acl_table_id == SAI_NULL_OBJECT_ID)
+    {
+        SWSS_LOG_ERROR(
+            "Cannot find corresponding ACL table for the entry %s",
+            sai_serialize_object_id(acl_entry_ids.front()).c_str());
+        return SAI_STATUS_FAILURE;
+    }
+    // Find MACsec SC
+    attr.id = SAI_MACSEC_SC_ATTR_FLOW_ID;
+    attr.value.oid = macsec_flow_id;
+    std::vector<sai_object_id_t> macsec_sc_ids;
+    findObjects(SAI_OBJECT_TYPE_MACSEC_SC, attr, macsec_sc_ids);
+    // One MACsec SC will only belong to one MACsec flow
+    // Meanwhile one MACsec flow will just belong to one port
+    if (macsec_sc_ids.empty())
+    {
+        SWSS_LOG_ERROR(
+            "Cannot find corresponding MACsec SC for the flow %s",
+            sai_serialize_object_id(macsec_flow_id).c_str());
+        return SAI_STATUS_FAILURE;
+    }
+    attrs.clear();
+    attrs.emplace_back();
+    attrs.back().id = SAI_MACSEC_SC_ATTR_MACSEC_DIRECTION;
+    CHECK_STATUS(
+        get(
+            SAI_OBJECT_TYPE_MACSEC_SC,
+            macsec_sc_ids.front(),
+            static_cast<uint32_t>(attrs.size()),
+            attrs.data()));
+    auto direction = attrs.back().value.s32;
+
+    // Find port
+    attrs.clear();
+    attrs.emplace_back();
+    if (direction == SAI_MACSEC_DIRECTION_EGRESS)
+    {
+        attrs.back().id = SAI_PORT_ATTR_EGRESS_MACSEC_ACL;
+    }
+    else
+    {
+        attrs.back().id = SAI_PORT_ATTR_INGRESS_MACSEC_ACL;
+    }
+    attrs.back().value.oid = acl_table_id;
+    std::vector<sai_object_id_t> port_ids;
+    findObjects(SAI_OBJECT_TYPE_PORT, attrs.back(), port_ids);
+    if (port_ids.empty())
+    {
+        SWSS_LOG_ERROR(
+            "Cannot find corresponding port for the ACL table %s",
+            sai_serialize_object_id(acl_table_id).c_str());
+        return SAI_STATUS_FAILURE;
+    }
+    else if (port_ids.size() > 1)
+    {
+        SWSS_LOG_ERROR(
+            "Duplicated ports for the ACL table %s",
+            sai_serialize_object_id(acl_table_id).c_str());
+        return SAI_STATUS_FAILURE;
+    }
+    port_id = port_ids.front();
+    return SAI_STATUS_SUCCESS;
```

Comment:
NOT a single empty line to distinguish whats going on here, just 1 big code blob :(